subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultImmutableAttributesContainer.java

```diff
@@ -29,10 +30,16 @@
 import java.util.Comparator;
 import java.util.LinkedHashMap;
 import java.util.Map;
+import java.util.Objects;
 import java.util.TreeMap;
 import java.util.concurrent.ConcurrentHashMap;
 
-final class DefaultImmutableAttributesContainer implements ImmutableAttributes, AttributeValue<Object> {
+/**
+ * Note: The fact that this type is both a <em>container</em> and a <em>value stored in a container</em> is strange.  This is
+ * probably something that ought to be addressed, this shouldn't implement {@link AttributeValue}, but it does it in order to
+ * build "linked lists" of immutable containers through concatenation to save memory.
+ */
+public final class DefaultImmutableAttributesContainer extends AbstractAttributeContainer implements ImmutableAttributes, AttributeValue<Object> {
```

Comment:
🤔 `DefaultImmutableAttributesContainer` inherits _no_ behavior from `AbstractAttributeContainer`. Having it implement `AbstractAttributeContainer` is more confusing than helpful to me. 

An `DefaultImmutableAttributesContainer` works much differently under the hood than any other attribute container. I would be suspicious if we did inherit any behavior from `AbstractAttributeContainer`

Reply:
Not true:
1. Most importantly, it still needs to check if an attribute is valid in the same way as the other implementations
2. It still needs to override and implement `equals` and `hashCode`
3. It also "benefits" from the implementation of `getAttributes()` in the abstract type (I need to remove this from `DIAC`)